File: thinking_and_programming_in_functional_style/code/clojure/exercise4/exercise.clj
;Given a number, determine the factorial of that number using tail recursion

(defn factorial [number]
  (defn factorial-impl [fact num]
    (if (== num 1)
      fact
      (recur (* fact num) (- num 1))))
  
  (factorial-impl 1 number))
  
(println (factorial 5))

File: thinking_and_programming_in_functional_style/code/scala/exercise4/exercise.scala
//Given a number, determine the factorial of that number using tail recursion

def factorial(number : Int) = {
  @scala.annotation.tailrec
  def factorialImpl(fact : Int, num : Int) : Int = {
    if (num == 1)
      fact
    else
      factorialImpl(fact * num, num - 1)
  }

  factorialImpl(1, number)  
}

println(factorial(5))